src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -0,0 +1,966 @@
+{
+    if (!isTuple(data_type))
+        return;
+
+    const auto * tuple_type = assert_cast<const DataTypeTuple *>(data_type.get());
+    auto nested_types = tuple_type->getElements();
+    transformInferredTypesIfNeededImpl<true>(nested_types, settings, json_info);
+    if (checkIfTypesAreEqual(nested_types))
+        data_type = std::make_shared<DataTypeArray>(nested_types.back());
+}
+
+
+template <bool is_json>
+static DataTypePtr tryInferDataTypeForSingleFieldImpl(ReadBuffer & buf, const FormatSettings & settings, JSONInferenceInfo * json_info);
+
+static bool tryInferDate(const std::string_view & field)
+{
+    ReadBufferFromString buf(field);
+    DayNum tmp;
+    return tryReadDateText(tmp, buf) && buf.eof();
+}
+
+static bool tryInferDateTime(const std::string_view & field, const FormatSettings & settings)
+{
+    if (field.empty())
+        return false;
+
+    ReadBufferFromString buf(field);
+    Float64 tmp_float;
+    /// Check if it's just a number, and if so, don't try to infer DateTime from it,
+    /// because we can interpret this number as a timestamp and it will lead to
+    /// inferring DateTime instead of simple Int64/Float64 in some cases.
+    if (tryReadFloatText(tmp_float, buf) && buf.eof())
+        return false;
+
+    buf.seek(0, SEEK_SET); /// Return position to the beginning
+    DateTime64 tmp;
+    switch (settings.date_time_input_format)
+    {
+        case FormatSettings::DateTimeInputFormat::Basic:
+            if (tryReadDateTime64Text(tmp, 9, buf) && buf.eof())
+                return true;
+            break;
+        case FormatSettings::DateTimeInputFormat::BestEffort:
+            if (tryParseDateTime64BestEffort(tmp, 9, buf, DateLUT::instance(), DateLUT::instance("UTC")) && buf.eof())
+                return true;
+            break;
+        case FormatSettings::DateTimeInputFormat::BestEffortUS:
+            if (tryParseDateTime64BestEffortUS(tmp, 9, buf, DateLUT::instance(), DateLUT::instance("UTC")) && buf.eof())
+                return true;
+            break;
+    }
+
+    return false;
+}
+
+DataTypePtr tryInferDateOrDateTimeFromString(const std::string_view & field, const FormatSettings & settings)
+{
+    if (settings.try_infer_dates && tryInferDate(field))
+        return std::make_shared<DataTypeDate>();
+
+    if (settings.try_infer_datetimes && tryInferDateTime(field, settings))
+        return std::make_shared<DataTypeDateTime64>(9);
+
+    return nullptr;
+}
+
+template <bool is_json>
+static DataTypePtr tryInferArray(ReadBuffer & buf, const FormatSettings & settings, JSONInferenceInfo * json_info)
+{
+    assertChar('[', buf);
+    skipWhitespaceIfAny(buf);
+
+    DataTypes nested_types;
+    bool first = true;
+    while (!buf.eof() && *buf.position() != ']')
+    {
+        if (!first)
+        {
+            /// Skip field delimiter between array elements.
+            skipWhitespaceIfAny(buf);
```

Comment:
you already skip it at the end of loop, and this can't be called in first iteration.